mlflow/genai/evaluation/utils.py

```diff
@@ -242,3 +245,86 @@ def eval_fn(
+
+    Scorer can return:
+    - A number, boolean, or string, a list of them.
+    - An Feedback object
+    - A list of Feedback objects
+
+    All of the above will be converted to a list of Feedback objects.
+    """
+    # None is a valid metric value, return an empty list
+    if value is None:
+        return []
+
+    # Primitives are valid metric values
+    if isinstance(value, (int, float, bool, str)):
+        return [
+            Feedback(
+                name=scorer_name,
+                source=make_code_type_assessment_source(scorer_name),
+                value=value,
+            )
+        ]
+
+    if isinstance(value, Feedback):
+        value.name = _get_custom_assessment_name(value, scorer_name)
+        return [value]
+
+    if isinstance(value, Collection):
+        assessments = []
+        for item in value:
+            if isinstance(item, Feedback):
+                # Scorer returns multiple assessments as a list.
+                item.name = _get_custom_assessment_name(item, scorer_name)
+                assessments.append(item)
+            else:
+                # If the item is not assessment, the list represents a single assessment
+                # value of list type. Convert it to a Feedback object.
+                assessments.append(
+                    Feedback(
+                        name=scorer_name,
+                        source=make_code_type_assessment_source(scorer_name),
+                        value=item,
+                    )
+                )
+        return assessments
+
+    raise MlflowException.invalid_parameter_value(
+        f"Got unsupported result from scorer '{scorer_name}'. "
+        f"Expected the metric value to be a number, or a boolean, or a string, "
+        "or an Feedback, or a list of Feedbacks. "
+        f"Got {value}.",
+    )
+
+
+def _get_custom_assessment_name(assessment: Feedback, scorer_name: str) -> str:
+    """Get the name of the custom assessment. Use assessment name if present and not a builtin judge
+    name, otherwise use the scorer name.
+
+    Args:
+        assessment: The assessment to get the name for.
+        scorer_name: The name of the scprer.
+    """
+    # If the user didn't provide a name, use the scorer name
+    if assessment.name == DEFAULT_FEEDBACK_NAME or (
+        assessment.metadata is not None
+        and assessment.metadata.get(USER_DEFINED_ASSESSMENT_NAME_KEY) == "false"
+    ):
+        return scorer_name
+    return assessment.name
+
+
+def make_code_type_assessment_source(scorer_name: str) -> AssessmentSource:
+    return AssessmentSource(source_type=AssessmentSourceType.CODE, source_id=scorer_name)
+
+
+def is_none_or_nan(value: Any) -> bool:
+    """Checks whether a value is None or NaN."""
+    # isinstance(value, float) check is needed to ensure that pd.isna is not called on an array.
+    return value is None or (isinstance(value, float) and pd.isna(value))
```

Comment:
```suggestion
    return value is None or (isinstance(value, float) and math.isnan(value))
```

can we use `math.isnan`?

Reply:
Sounds good. Btw I was a bit surprised that this function doesn't support `pd.NA` (it returns false). Will add a note about it.